### PR TITLE
Add customer address API

### LINKED ISSUE #181: Add missing Customer Address API
Should be consistent with: https://github.com/Shopify/mobile-buy-sdk-android/blob/version-2/MobileBuy/buy/src/main/java/com/shopify/buy/dataprovider/AddressService.java

@gabrieloc @dbart01 @davidmuzi

### PR DESCRIPTION
**Pull request by dbart01**
### What this does
- adds API for creating, fetching, updating & deleting customer addresses

Fixes #181 
Depends on https://github.com/Shopify/shopify/issues/73424#issuecomment-222145874

@bgulanowski @davidmuzi

### CODE REVIEW

File: Mobile Buy SDK/Mobile Buy SDK/Operations/BUYRequestOperation.m [hunk 1]
```diff
@@ -169,14 +168,24 @@ - (NSURLSessionDataTask *)request:(NSURLRequest *)request completion:(BUYRequest
 			json = [NSJSONSerialization JSONObjectWithData:data options:0 error:nil];
 		}
 		
-		NSHTTPURLResponse *httpResponse = (id)response;
-		if (!httpResponse.successful && !error) {
```
Comment (davidmuzi) [inline Mobile Buy SDK/Mobile Buy SDK/Operations/BUYRequestOperation.m]: Should the method be removed too?

File: Mobile Buy SDK/Mobile Buy SDK/Data/BUYClient.h [hunk 2]
```diff
@@ -147,15 +142,6 @@ NS_ASSUME_NONNULL_BEGIN
...
 @property (strong, nonatomic, nullable) NSString *customerToken;
 
-#pragma mark - Deprecated methods
-
-/**
- *  Enable Apple Pay by calling this method with the Merchant ID provided via Apple Pay setup in the Mobile SDK Channel on Shopify Admin
- *
- *  @param merchantId The Merchant ID generated on Shopify Admin
- */
-- (void)enableApplePayWithMerchantId:(NSString *)merchantId NS_DEPRECATED_IOS(8_0, 9_0, "Set the merchantId on a `BUYViewController` subclass instead");
```
Comment (davidmuzi) [inline Mobile Buy SDK/Mobile Buy SDK/Data/BUYClient.h]: 👍

File: Mobile Buy SDK/Mobile Buy SDK Tests/BUYClientTest_Customer.m [hunk 3]
```diff
@@ -148,9 +209,204 @@ - (void)testCustomerLogin
...
 	
-	[self waitForExpectationsWithTimeout:10 handler:^(NSError * _Nullable error) {
+	[self waitForExpectationsWithTimeout:10 handler:nil];
+}
+
+#pragma mark - Address -
+
+- (void)testGetAddresses
+{
+	//	[OHHTTPStubs stubRequestsPassingTest:^BOOL(NSURLRequest * _Nonnull request) {
```
Comment (davidmuzi) [inline Mobile Buy SDK/Mobile Buy SDK Tests/BUYClientTest_Customer.m]: why are these left out?

File: Mobile Buy SDK/Mobile Buy SDK Tests/mocked_responses.json [hunk 4]
```diff
@@ -123,5 +123,9 @@
...
..."province\":null,\"country\":\"Bouvet Island\",\"zip\":\"24124124\",\"phone\":\"123124124\",\"name\":\"Fast Add\",\"province_code\":null,\"country_code\":\"BV\",\"country_name\":\"Bouvet Island\",\"default\":true},\"verified_email\":true,\"accepts_marketing\":false,\"first_name\":\"Fast\",\"last_name\":\"Add\",\"orders_count\":9,\"total_spent\":\"375.00\",\"created_at\":\"2016-02-16T14:42:24-05:00\",\"updated_at\":\"2016-03-10T16:34:04-05:00\",\"state\":\"enabled\",\"last_order_id\":2566266118,\"last_order_name\":\"#1137\",\"addresses\":[{\"id\":2785988486,\"first_name\":\"AA\",\"last_name\":\"A\",\"phone\":\"\",\"company\":\"\",\"address1\":\"Assiniboine Road\",\"address2\":\"\",\"city\":\"Toronto\",\"province\":\"Ontario\",\"province_code\":\"ON\",\"country\":\"Canada\",\"country_code\":\"CA\",\"zip\":\"M3J1E1\"},{\"id\":2839567814,\"first_name\":\"Fast\",\"last_name\":\"Add\",\"phone\":\"123124124\",\"company\":\"Adidas\",\"address1\":\"Sass\",\"address2\":\"12\",\"city\":\"Qsdasd\",\"province\":null,\"province_code\":null,\"country\":\"Bouvet Island\",\"country_code\":\"BV\",\"zip\":\"24124124\"}],\"multipass_identifier\":null,\"tax_exempt\":false}}","code":200,"message":"OK"},
```
Comment (davidmuzi) [inline Mobile Buy SDK/Mobile Buy SDK Tests/mocked_responses.json]: lets not use this company name

File: Mobile Buy SDK/Mobile Buy SDK/Data/BUYClient+Address.m [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+{
+	BUYAssert(address.identifier, @"Failed to update address. Address must have a valid identifier.");
+	
+	NSURL *route = [self urlForCustomersAddressWithID:customerID addressID:address.identifier];
+	return [self putRequestForURL:route object:@{ @"address" : address.JSONDictionary } completionHandler:^(NSDictionary *json, NSHTTPURLResponse *response, NSError *error) {
+		BUYAddress *address = nil;
+		if (json && !error) {
+			address = [self.modelManager insertAddressWithJSONDictionary:json[@"address"]];
+		}
+		block(address, error);
```
Comment (davidmuzi) [inline Mobile Buy SDK/Mobile Buy SDK/Data/BUYClient+Address.m]: If the callbacks are non-nullable we should mark them as such, or check for null before calling

File: Mobile Buy SDK/Mobile Buy SDK Tests/mocked_responses.json (hunk 4 above)
Comment (dbart01) [inline Mobile Buy SDK/Mobile Buy SDK Tests/mocked_responses.json]: That was in there from before, it really shouldn't be part of this diff but I can change it.

File: Mobile Buy SDK/Mobile Buy SDK/Data/BUYClient+Address.m (hunk 5 above)
Comment (dbart01) [inline Mobile Buy SDK/Mobile Buy SDK/Data/BUYClient+Address.m]: It is marked. Using: `NS_ASSUME_NONNULL_BEGIN` in the header.

File: Mobile Buy SDK/Mobile Buy SDK Tests/BUYClientTest_Customer.m (hunk 3 above)
Comment (dbart01) [inline Mobile Buy SDK/Mobile Buy SDK Tests/BUYClientTest_Customer.m]: My mistake, will update.

File: Mobile Buy SDK/Mobile Buy SDK/Operations/BUYRequestOperation.m (hunk 1 above)
Comment (dbart01) [inline Mobile Buy SDK/Mobile Buy SDK/Operations/BUYRequestOperation.m]: I'm not sure I understand which method should be removed?

Comment (davidmuzi) [inline Mobile Buy SDK/Mobile Buy SDK/Operations/BUYRequestOperation.m]:
```
- (BOOL)successful
{
	return ((NSUInteger)(self.statusCode / 100)) == 2;
}
```

On second thought, its still used, so it should be called from `errorFromJSON`

**Comment by bgulanowski** [comment]
I don't have any suggestions. Looks great.

**Pull request merged**